### DISCUSSION TITLE
Extending AppStorage to support other types (Namely Codable)

### DISCUSSION THREAD

**Discussion by brzzdev** [pointfreeco/swift-sharing#48] [Ideas]
I just created [a PR](https://github.com/pointfreeco/swift-sharing/pull/47) and didn't realise @rafalwojcik had already [created one](https://github.com/pointfreeco/swift-sharing/pull/18) which exposes less of the internals and is all around simpler 👍🏻.

I do see your arguments for keeping parity with the constraints of `@AppStorage`, but I also think there's a huge need for allowing us some sort of custom override.

### Why:
**- Legacy usecases**
Many of us already have dozens of `Codable` values in `UserDefaults` across multiple targets and apps. We could migrate them over to `fileStorage`, but this is quite time-consuming, prone to errors and lost data, and…

**- File storage privacy**
Obviously, neither are secure, and you can view both the `UserDefaults` `.plist` and the files stored by `fileStorage`, but `UserDefaults` is a single file where Data fields are a hex that has to be decoded. `fileStorage` creates a file for each object that plainly shows the object. I would argue that `UserDefaults` provides a little more security through obscurity out of the box.

**- Standard practice**
If you look on GitHub at any other popular library that writes to UserDefaults, I struggle to find one that doesn't offer `Codable` support. [Foil](https://github.com/jessesquires/Foil) is the only one I could find that warned against it, but even that supports it out of the box.

I think a solution like @rafalwojcik's that makes `AppStorageKey` flexible enough for everyone's needs, while still allowing you guys to caution against it and recommend `fileStorage` for `Codable` objects going forward, is a win-win.

**Comment by stephencelis** [comment]
Hi @brzzdev,

Thanks for starting a discussion.

I think it's important to reiterate that the library supports defining custom keys outside the core library. So you are free to create a custom `CodableAppStorageKey` key for this use case, and you could even package it up for folks to use. You might even be able to leverage `AppStorageKey<Data>` under the hood if you're worried about maintaining that functionality all over again.

I think the core library, though, should probably remain the way it is to discourage storing codable values in user defaults, as Apple does by convention, and as Foil does by documentation. The library doesn't _prevent_ you from achieving the functionality you're asking for, it just doesn't come with that functionality baked into it.

> **- File storage privacy** Obviously, neither are secure, and you can view both the `UserDefaults` `.plist` and the files stored by `fileStorage`, but `UserDefaults` is a single file where Data fields are a hex that has to be decoded. `fileStorage` creates a file for each object that plainly shows the object. I would argue that `UserDefaults` provides a little more security through obscurity out of the box.

I'm not sure I get this argument. Can you explain what could go wrong here?

**Reply by brzzdev** [reply]
Hi @stephencelis,

Thanks for the prompt follow up.

> I think it's important to reiterate that the library supports defining custom keys outside the core library. So you are free to create a custom CodableAppStorageKey key for this use case, and you could even package it up for folks to use.

My issue there is that I want 99.9% of the functionality of your existing and maintained `AppStorageKey`. I just want a little extra on top.
The solution we've gone with since `Shared` first launch is just reimplementing `AppStorageKey.swift` and adding `Codable` and custom store stuff. (I saw you merged support for custom stores on the key level. Nice.)
This leads to us chasing a moving target as you improve `AppStorageKey` and fix bugs. Someone has to track the changes to that one file to ensure we're not missing something.

> You might even be able to leverage AppStorageKey<Data> under the hood if you're worried about maintaining that functionality all over again.

I'd love to be able to do that, but as far as I can tell, all the `fileprivate` inits prevent that. That's why my PR was focused around just making a public init, and some types and protocols that you use public.

> I'm not sure I get this argument. Can you explain what could go wrong here?

Off the top of my head, say we store an `Experiment` object that defines if a user is in group A or B. With the fileStorage approach, the user can easily see this in plain text and put themselves in a different group. They could theoretically do the same with UserDefaults but it requires a degree more technical knowledge.

**Reply by acosmicflamingo** [reply]
I'd imagine that Point-Free would want to avoid a situation where they give the people what they want, people abuse it, bugs arise, and same people come to them asking why X is not working as expected.

Maybe Point-Free can include some of this API in some _spi import label dubbed "unstable" so when things do break, Point-Free can say "It's called unstable for a reason" LOL

**Reply by rafalwojcik** [reply]
Initially, I was a bit bummed when my changes were rejected, but after giving it some thought, I think it’s actually a good idea. The only real advantage of AppStorage is that it can automatically sync with other apps without any extra work. But here’s the catch: this benefit could cause serious problems if different versions of Codable objects are stored in different apps. And guess what? Apple doesn’t have a clear way to handle conflicts between UserDefauls objects. So, I think Apple’s approach of using AppStorage in that case is a good move.

**Reply by stephencelis** [reply]
> I'd love to be able to do that, but as far as I can tell, all the fileprivate inits prevent that. That's why my PR was focused around just making a public init, and some types and protocols that you use public.

The `.appStorage` static is public and I think that's all you need to define a wrapper. If you hit any stumbling blocks, though, let me know!

**Reply by myihsan** [reply]
> The .appStorage static is public and I think that's all you need to define a wrapper.

I have successfully defined a wrapper key.
However, it breaks after updating to 2.* since `LoadContinuation.init` is `package`.
As a workaround, I am thinking of defining a wrapper for the value and conforming to `RawRepresentable`.

Am I missing something?

**Reply by kgrigsby59** [reply]
I've run into the same problem.

**Reply by dadalar** [reply]
Same here

**Reply by stephencelis** [reply]
As of 2.1.1 and https://github.com/pointfreeco/swift-sharing/pull/81, you can initialize the continuation types to support this functionality.

**Reply by pyrtsa** [reply]
Wow this was timely, was working on the same problem and just about to ask why not turn the initialisers `public` the very minute you published 2.1.1. :tada:
